### ISSUE TITLE
Question:  How to attach analyzers and code fixes to a project?

### ISSUE THREAD

**Issue by shaggygi** [DotNetAnalyzers/StyleCopAnalyzers#1984]
This is more of a Roslyn question, but thought this team might know since I want to try with StyelCop.Analyzers and code fixes.

How would you use a set of analyzers from rule set with a project in Visual Studio and programmatically check existing warnings along with code fixes?  Is there a way to use the new C# Interactive or Console App instead of manually fixing in Visual Studio?  Then go a little further and automatically fix the code and save back to file(s) similar to clicking the Lightbulb > Fix all occurrences in...

As an example, how could you check for all violations for a simple analyzer like SA1028 (trailing spaces)?

Any help is appreciated.

**Comment by Parsavali** [comment]
This is my question as well. I want to know is there any way to apply code fixes using a console application automatically(not just using the Lightbulb in VS)?

**Comment by vweijsters** [comment]
I think that our [StyleCopTester](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/tree/master/StyleCop.Analyzers/StyleCopTester) console application could be a good starting point for that. Have a look and let us know if you have additional questions.

**Comment by shaggygi** [comment]
Thanks! This is exactly what I needed to get started.  Understanding this is another question not specifically related to this project, but do you know of any examples how to 1.) programmatically add the latest StyleCop.Analyzer NuGet package to a project and 2.) programmatically set a rule set file for a project?

Thanks again.

**Comment by Parsavali** [comment]
thanks @vweijsters for your answer. I looked at the StyleCopTester source code, that is what I needed but one developer-side question. As I understood, foreach diagnostics it will prepare a list of codefixes(List<CodeAction> fixes). Am I right? 
If that is so, Is that a good idea to apply a code fix in CI and save back to the disk automatically? Because there is not just one code fix but list of possible code fixes as I understood. What is your idea?

**Comment by Parsavali** [comment]
@shaggygi 
1) just <Package> tag to the packages.config file in your source project then you have to restore packages programmatically so that it installs stylecopAnalyzer to the packages folder.
2) you can create your own custom .ruleset file in your project and reference to it in the .csproj project file(https://msdn.microsoft.com/en-us/library/dd264974.aspx)

**Comment by shaggygi** [comment]
@Parsavali thanks

**Comment by vweijsters** [comment]
@Parsavali It is possible to have multiple code fixes presented for a diagnostic, but this will not happen often (at least not for StyleCop Analyzers), as code fixes are specific for a diagnostic. You will have to examine each code fix that is offered and pick the one that is appropriate for your solution, or skip the code fix all together if there is no clear resolution. 

Note that not all diagnostics will always have a code fix available. We try to provide code fixes as much as possible, but sometimes there is not enough information available to provide a fix (for example: summary fields in documentation comments).

**Comment by Parsavali** [comment]
thanks @vweijsters for the complete answer. Is there any documentation regarding StyleCop Analyzers diagnostics and corresponding code fixes? Or I have to run the code and see what kind of code fixes exist for each diagnostics?

**Comment by vweijsters** [comment]
If you inspect the source code of the [code fixes](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/tree/master/StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes), each code fix will list the diagnostics that it supports (the `FixableDiagnosticIds` property). Look in the `RegisterCodeFixesAsync` method override to see what code fix is offered (you can determine here if multiple code fixes will be offered as well).

**Comment by Parsavali** [comment]
Thank you so much @vweijsters. As you said I went through the codes and got the idea;). I changed the code in a way so that code fixes can be assigned to corresponding textspans and change the code and save back to disk but there is one problem here.
Imagine we have a piece of code like this in our project:
```javascript
private static void Main()
    {
        //This comment is for test
	int r = CalculatorLib.Calculator.Add(3, 4);
	Console.WriteLine("Result is:{0}",r);
	//This comment is for test
        Console.ReadLine();
    }
```
For this piece of code there are two violations of SA1005 which is about adding a space affter each comments slashes which should be like this after code fixes:
```javascript
private static void Main()
    {
        // This comment is for test
	int r = CalculatorLib.Calculator.Add(3, 4);
	Console.WriteLine("Result is:{0}",r);
	// This comment is for test
        Console.ReadLine();
    }
```
The problem is that in the GetTransformedDocumentAsync method(For example in SA1005CodeFixProvider.cs file) when it wants to call GetTextChange method to apply code fix, it passes location.SourceSpan which is the start and end location of the violation(here for example for the first comment, location.SourceSpan = {[807..832)} and for the second comment location.SourceSpan = {[888..913)). When I apply codefix for the first comment it is completely ok and put one space for the first comment and save back the file but the problem here is that the location of the second violation is [888..913] but that should be changed after adding one space for the first comment and it should take affect on this new possition [889..914] not the previous one.
That is the problem I faced and I this was just for one violation scenario as others have the same problem for sure. 
Do you have any idea how can I cop with that?

**Comment by sharwell** [comment]
If you take the [result of `GetOperationsAsync` here](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/f4e3909012d52368cbd9064b92da6293075861d1/StyleCop.Analyzers/StyleCopTester/Program.cs#L198) and call `ApplyChanges`, the changes will be written back to files on disk. The code fix and fix all already ensure that overlapping source spans do not cause problems during this operation.
